### ISSUE TITLE
Bazel 7 crashes if coverage report is a tree artifact

### ISSUE THREAD

**Issue by thii** [bazelbuild/bazel#22069]
### Description of the bug:

When using a custom coverage report generator to generate a coverage report (which is hardcoded at `bazel-out/_coverage/_coverage_report.dat`), Bazel 7 crashes if this is a directory (works fine in 6.x).

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

Create a Bazel workspace with:

main.c
```
int main() { return 0; }
```

report_gen.sh
```
#!/bin/bash

rm -rf bazel-out/_coverage/_coverage_report.dat
mkdir bazel-out/_coverage/_coverage_report.dat
```

BUILD
```
cc_test(
    name = "main",
    srcs = ["main.c"],
)

sh_binary(
    name = "report_gen",
    srcs = ["report_gen.sh"],
)
```

Then run:
```
bazel coverage //... --coverage_report_generator=//:report_gen --combined_report=lcov
```
```
Starting local Bazel server and connecting to it...
INFO: Using default value for --instrumentation_filter: "^//".
INFO: Override the above default with --instrumentation_filter
INFO: Analyzed 2 targets (77 packages loaded, 1157 targets configured).
INFO: LCOV coverage report is located at /home/user/.cache/bazel/_bazel_user/7d8c4501c48b39c8c25928d54354d56a/execroot/_main/bazel-out/_coverage/_coverage_report.dat
 and execpath is bazel-out/_coverage/_coverage_report.dat
[27 / 28] 1 / 1 tests;  1 action; last test: //:main
    Coverage report generation; 0s linux-sandbox
FATAL: bazel crashed due to an internal error. Printing stack trace:
java.lang.IllegalStateException: java.lang.RuntimeException: Unrecoverable error while evaluating node 'UnshareableActionLookupData{actionLookupKey=CoverageReportKeySingleton, actionIndex=1}' (requested by nodes )
        at com.google.devtools.build.lib.skyframe.SkyframeExecutor.evaluateSkyKeys(SkyframeExecutor.java:1858)
        at com.google.devtools.build.lib.skyframe.SkyframeBuildView.analyzeAndExecuteTargets(SkyframeBuildView.java:738)
        at com.google.devtools.build.lib.analysis.BuildView.update(BuildView.java:292)
        at com.google.devtools.build.lib.buildtool.AnalysisAndExecutionPhaseRunner.runAnalysisAndExecutionPhase(AnalysisAndExecutionPhaseRunner.java:241)
        at com.google.devtools.build.lib.buildtool.AnalysisAndExecutionPhaseRunner.execute(AnalysisAndExecutionPhaseRunner.java:139)
        at com.google.devtools.build.lib.buildtool.BuildTool.buildTargetsWithMergedAnalysisExecution(BuildTool.java:305)
        at com.google.devtools.build.lib.buildtool.BuildTool.buildTargets(BuildTool.java:173)
        at com.google.devtools.build.lib.buildtool.BuildTool.processRequest(BuildTool.java:510)
        at com.google.devtools.build.lib.buildtool.BuildTool.processRequest(BuildTool.java:478)
        at com.google.devtools.build.lib.runtime.commands.TestCommand.doTest(TestCommand.java:163)
        at com.google.devtools.build.lib.runtime.commands.TestCommand.exec(TestCommand.java:116)
        at com.google.devtools.build.lib.runtime.BlazeCommandDispatcher.execExclusively(BlazeCommandDispatcher.java:664)
        at com.google.devtools.build.lib.runtime.BlazeCommandDispatcher.exec(BlazeCommandDispatcher.java:244)
        at com.google.devtools.build.lib.server.GrpcServerImpl.executeCommand(GrpcServerImpl.java:573)
        at com.google.devtools.build.lib.server.GrpcServerImpl.lambda$run$1(GrpcServerImpl.java:644)
        at io.grpc.Context$1.run(Context.java:566)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.base/java.lang.Thread.run(Unknown Source)
Caused by: java.lang.RuntimeException: Unrecoverable error while evaluating node 'UnshareableActionLookupData{actionLookupKey=CoverageReportKeySingleton, actionIndex=1}' (requested by nodes )
        at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:550)
        at com.google.devtools.build.lib.concurrent.AbstractQueueVisitor$WrappedRunnable.run(AbstractQueueVisitor.java:414)
        at java.base/java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinPool.scan(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(Unknown Source)
Caused by: java.lang.NullPointerException: Cannot invoke "com.google.devtools.build.lib.cmdline.Label.toString()" because the return value of "com.google.devtools.build.lib.actions.ActionOwner.getLabel()" is null
        at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor.checkForUnsoundDirectoryOutput(SkyframeActionExecutor.java:1670)
        at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor.checkOutputs(SkyframeActionExecutor.java:1551)
        at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$ActionRunner.actuallyCompleteAction(SkyframeActionExecutor.java:1221)
        at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$ActionRunner.executeAction(SkyframeActionExecutor.java:1170)
        at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$ActionRunner.run(SkyframeActionExecutor.java:1061)
        at com.google.devtools.build.lib.skyframe.ActionExecutionState.runStateMachine(ActionExecutionState.java:165)
        at com.google.devtools.build.lib.skyframe.ActionExecutionState.getResultOrDependOnFuture(ActionExecutionState.java:94)
        at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor.executeAction(SkyframeActionExecutor.java:558)
        at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.checkCacheAndExecuteIfNeeded(ActionExecutionFunction.java:859)
        at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.computeInternal(ActionExecutionFunction.java:333)
        at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.compute(ActionExecutionFunction.java:171)
        at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:461)
        ... 7 more
```

### Which operating system are you running Bazel on?

Linux

### What is the output of `bazel info release`?

7.1.1 release

**Comment by tjgq** [comment]
Thanks for the repro instructions. Crashing is definitely not acceptable and I'm preparing a fix.

Regarding emitting `coverage.dat` as a directory: Bazel needs to know whether an artifact is a file or a directory before its contents are produced, and `coverage.dat` is declared as a file. Before Bazel 7, producing a directory where a file is expected was tolerated, but broken in subtle ways. Bazel 7 forbids it by default (`--incompatible_disallow_unsound_directory_outputs`) and Bazel 8 will not allow it at all. So while fixing the crash should allow you to disable `--incompatible_disallow_unsound_directory_outputs` and build successfully, this will only work until Bazel 8.

If producing `coverage.dat` as a directory after Bazel 8 is a requirement, please open a separate feature request. (Note that this would necessarily require adding some sort of new API, since a single artifact can't sometimes be a file and sometimes a directory, and `coverage.dat` is already a file in most cases.)

**Comment by fmeum** [comment]
@bazel-io fork 7.2.0

**Comment by iancha1992** [comment]
A fix for this issue has been included in [Bazel 7.2.0 RC1](https://github.com/bazelbuild/bazel/releases/tag/7.2.0rc1). Please test out the release candidate and report any issues as soon as possible.
If you're using Bazelisk, you can point to the latest RC by setting USE_BAZEL_VERSION=7.2.0rc1. Thanks!
